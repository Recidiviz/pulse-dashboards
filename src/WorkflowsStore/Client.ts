// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { entries, makeObservable, remove, set, when } from "mobx";
import { format as formatPhone } from "phone-fns";

import {
  trackClientProfileViewed,
  trackOpportunityPreviewed,
  trackProfileOpportunityClicked,
  trackReferralFormPrinted,
  trackReferralFormViewed,
  trackSetOpportunityStatus,
  trackSurfacedInList,
} from "../analytics";
import {
  ClientRecord,
  FullName,
  SpecialConditionCode,
  updateOpportunityCompleted,
  updateOpportunityDenial,
} from "../firestore";
import type { RootStore } from "../RootStore";
import {
  CompliantReportingOpportunity,
  EarlyTerminationOpportunity,
  EarnedDischargeOpportunity,
  LSUOpportunity,
  Opportunity,
  OPPORTUNITY_TYPES,
  OpportunityType,
  PastFTRDOpportunity,
} from "./Opportunity";
import { optionalFieldToDate } from "./utils";
import { OTHER_KEY } from "./WorkflowsStore";

export const UNKNOWN = "Unknown" as const;

const SUPERVISION_LEVEL_MAP: Record<string, string> = {
  // TODO(#2287): remove these levels derived from TN raw data
  "STANDARD: MEDIUM": "Medium",
  "STANDARD: MINIMUM": "Minimum",
};

type ClientDetailsCopy = {
  supervisionStartDate?: {
    tooltip: string;
  };
};

export const CLIENT_DETAILS_COPY: Record<string, ClientDetailsCopy> = {
  US_ND: {
    supervisionStartDate: {
      tooltip:
        "This date is included in the filled form as the date that the Defendant was sentenced; in rare cases, the sentencing date and supervision start date may differ. Double check the date in the form to ensure that the sentencing date and supervision start date match in this case.",
    },
  },
};

type OpportunityConfig = {
  flag: `${OpportunityType}Eligible`;
  OpportunityClass: {
    new (client: Client): Opportunity;
  };
};

const OPPORTUNITY_CREATION_MAPPING: Record<
  OpportunityType,
  OpportunityConfig
> = {
  compliantReporting: {
    flag: "compliantReportingEligible",
    OpportunityClass: CompliantReportingOpportunity,
  },
  earlyTermination: {
    flag: "earlyTerminationEligible",
    OpportunityClass: EarlyTerminationOpportunity,
  },
  earnedDischarge: {
    flag: "earnedDischargeEligible",
    OpportunityClass: EarnedDischargeOpportunity,
  },
  LSU: { flag: "LSUEligible", OpportunityClass: LSUOpportunity },
  pastFTRD: { flag: "pastFTRDEligible", OpportunityClass: PastFTRDOpportunity },
};

type OpportunityMapping = {
  earlyTermination?: EarlyTerminationOpportunity;
  compliantReporting?: CompliantReportingOpportunity;
  earnedDischarge?: EarnedDischargeOpportunity;
  LSU?: LSUOpportunity;
  pastFTRD?: PastFTRDOpportunity;
};

export class Client {
  rootStore: RootStore;

  recordId!: string;

  id!: string;

  record!: ClientRecord;

  pseudonymizedId!: string;

  stateCode!: string;

  fullName!: FullName;

  officerId!: string;

  supervisionType!: string;

  supervisionLevelStart?: Date;

  address!: string;

  private rawPhoneNumber?: string;

  expirationDate?: Date;

  supervisionStartDate?: Date;

  currentBalance?: number;

  lastPaymentAmount?: number;

  lastPaymentDate?: Date;

  probationSpecialConditions?: string[];

  paroleSpecialConditions?: SpecialConditionCode[];

  formIsPrinting = false;

  opportunities: OpportunityMapping = {};

  constructor(record: ClientRecord, rootStore: RootStore) {
    makeObservable(this, {
      opportunities: true,
      record: true,
      currentUserEmail: true,
      formIsPrinting: true,
      opportunitiesAlmostEligible: true,
      opportunitiesEligible: true,
      printReferralForm: true,
      setFormIsPrinting: true,
      supervisionLevel: true,
      updateRecord: true,
    });

    this.rootStore = rootStore;

    this.updateRecord(record);
  }

  updateRecord(record: ClientRecord): void {
    this.recordId = record.recordId;
    this.record = record;
    this.id = record.personExternalId;
    this.pseudonymizedId = record.pseudonymizedId;
    this.stateCode = record.stateCode.toUpperCase();
    this.fullName = record.personName;
    this.officerId = record.officerId;
    this.supervisionType = record.supervisionType;
    this.supervisionLevelStart = optionalFieldToDate(
      record.supervisionLevelStart
    );
    this.address = record.address || UNKNOWN;
    this.rawPhoneNumber = record.phoneNumber;
    this.expirationDate = optionalFieldToDate(record.expirationDate);
    this.currentBalance = record.currentBalance;
    this.lastPaymentDate = optionalFieldToDate(record.lastPaymentDate);
    this.lastPaymentAmount = record.lastPaymentAmount;
    this.probationSpecialConditions = record.specialConditions;
    this.paroleSpecialConditions = record.boardConditions ?? [];
    this.supervisionStartDate = optionalFieldToDate(
      record.supervisionStartDate
    );

    OPPORTUNITY_TYPES.forEach((t) => {
      const { flag, OpportunityClass } = OPPORTUNITY_CREATION_MAPPING[t];
      if (record[flag]) {
        if (!this.opportunities[t]) {
          set(this.opportunities, t, new OpportunityClass(this));
        }
      } else {
        remove(this.opportunities, t);
      }
    });
  }

  get displayName(): string {
    return [this.fullName.givenNames, this.fullName.surname]
      .filter((n) => Boolean(n))
      .join(" ");
  }

  get supervisionLevel(): string {
    const levelFromFilters = this.rootStore.workflowsStore.supervisionLevels.find(
      (opt) => opt.value === this.record.supervisionLevel
    )?.label;
    // some levels are mapped locally for historical reasons (e.g., to patch gaps in ingest or handle raw data)
    const locallyMappedLevel =
      SUPERVISION_LEVEL_MAP[this.record.supervisionLevel ?? ""];
    return (
      levelFromFilters ??
      locallyMappedLevel ??
      // TODO(#2287): remove TN-specific override once data is migrated;
      // this prevents a sudden proliferation of "Unknown" labels in TN in the interim
      (this.stateCode === "US_TN" ? "" : UNKNOWN)
    );
  }

  get formattedPhoneNumber(): string | undefined {
    return this.rawPhoneNumber
      ? formatPhone("(NNN) NNN-NNNN", this.rawPhoneNumber)
      : undefined;
  }

  get phoneNumber(): string {
    return this.formattedPhoneNumber || UNKNOWN;
  }

  get officerDistrict(): string | undefined {
    const officer = this.rootStore.workflowsStore?.availableOfficers.find(
      (o) => o.id === this.officerId
    );
    return officer?.district;
  }

  get opportunitiesEligible(): OpportunityMapping {
    return entries(this.opportunities).reduce((opportunities, [key, opp]) => {
      if (opp?.isHydrated && !opp.almostEligible) {
        return { ...opportunities, [key as OpportunityType]: opp };
      }
      return opportunities;
    }, {} as OpportunityMapping);
  }

  get opportunitiesAlmostEligible(): OpportunityMapping {
    return entries(this.opportunities).reduce((opportunities, [key, opp]) => {
      if (opp?.isHydrated && opp.almostEligible) {
        return { ...opportunities, [key as OpportunityType]: opp };
      }
      return opportunities;
    }, {} as OpportunityMapping);
  }

  get currentUserEmail(): string | null | undefined {
    return this.rootStore.workflowsStore.user?.info.email;
  }

  get currentUserName(): string | null | undefined {
    return this.rootStore.workflowsStore.user?.info.email;
  }

  async setOpportunityDenialReasons(
    reasons: string[],
    opportunityType: OpportunityType
  ): Promise<void> {
    if (this.currentUserEmail) {
      // clear irrelevant "other" text if necessary
      const deletions = reasons.includes(OTHER_KEY)
        ? undefined
        : { otherReason: true };

      await updateOpportunityDenial(
        this.currentUserEmail,
        this.recordId,
        { reasons },
        opportunityType,
        deletions
      );

      await updateOpportunityCompleted(
        this.currentUserEmail,
        this.recordId,
        opportunityType,
        true
      );

      if (reasons.length) {
        trackSetOpportunityStatus({
          clientId: this.pseudonymizedId,
          status: "DENIED",
          opportunityType,
          deniedReasons: reasons,
        });
      } else {
        trackSetOpportunityStatus({
          clientId: this.pseudonymizedId,
          status: "IN_PROGRESS",
          opportunityType,
        });
      }
    }
  }

  async setOpportunityOtherReason(
    opportunityType: OpportunityType,
    otherReason?: string
  ): Promise<void> {
    if (this.currentUserEmail) {
      await updateOpportunityDenial(
        this.currentUserEmail,
        this.recordId,
        {
          otherReason,
        },
        opportunityType
      );
    }
  }

  setFormIsPrinting(value: boolean): void {
    this.formIsPrinting = value;
  }

  printReferralForm(opportunityType: OpportunityType): void {
    const opportunity = this.opportunities[opportunityType];
    if (this.currentUserEmail && opportunity?.isHydrated) {
      if (opportunity.reviewStatus !== "DENIED") {
        updateOpportunityCompleted(
          this.currentUserEmail,
          this.recordId,
          opportunityType
        );
        if (opportunity.reviewStatus !== "COMPLETED") {
          trackSetOpportunityStatus({
            clientId: this.pseudonymizedId,
            status: "COMPLETED",
            opportunityType,
          });
        }
      }

      this.setFormIsPrinting(true);
      trackReferralFormPrinted({
        clientId: this.pseudonymizedId,
        opportunityType,
      });
    }
  }

  async trackFormViewed(formType: OpportunityType): Promise<void> {
    await when(() => this.opportunities[formType]?.isLoading === false);

    trackReferralFormViewed({
      clientId: this.pseudonymizedId,
      opportunityType: formType,
    });
  }

  async trackListViewed(listType: OpportunityType): Promise<void> {
    await when(() => this.opportunities[listType]?.isLoading === false);

    trackSurfacedInList({
      clientId: this.pseudonymizedId,
      opportunityType: listType,
    });
  }

  trackProfileViewed(): void {
    trackClientProfileViewed({ clientId: this.pseudonymizedId });
  }

  trackProfileOpportunityClicked(opportunityType: OpportunityType): void {
    trackProfileOpportunityClicked({
      clientId: this.pseudonymizedId,
      opportunityType,
    });
  }

  async trackOpportunityPreviewed(
    opportunityType: OpportunityType
  ): Promise<void> {
    await when(() => this.opportunities[opportunityType]?.isLoading === false);

    trackOpportunityPreviewed({
      clientId: this.pseudonymizedId,
      opportunityType,
    });
  }

  get detailsCopy(): ClientDetailsCopy {
    return CLIENT_DETAILS_COPY[this.stateCode];
  }
}
