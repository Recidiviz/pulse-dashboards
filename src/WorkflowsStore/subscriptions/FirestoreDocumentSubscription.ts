// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import * as Sentry from "@sentry/react";
import {
  DocumentData,
  DocumentReference,
  DocumentSnapshot,
  onSnapshot,
  Unsubscribe,
} from "firebase/firestore";
import { action, makeObservable, observable, runInAction } from "mobx";

import { FeatureGateError } from "../../errors";
import { castToError } from "../../utils/castToError";
import { FirestoreSubscription } from "./FirestoreSubscription";
import {
  DocumentSubscription,
  TransformFunction,
  UpdateFunction,
  ValidateFunction,
} from "./types";
import {
  defaultTransformFunction,
  defaultUpdateFunction,
  defaultValidateFunction,
} from "./utils";

export abstract class FirestoreDocumentSubscription<
    DataFormat extends DocumentData = DocumentData
  >
  extends FirestoreSubscription<DataFormat | undefined>
  implements DocumentSubscription<DataFormat>
{
  abstract readonly dataSource: DocumentReference;

  constructor(
    // default passes through raw record, assuming it already conforms to the desired format
    protected transformRecord: TransformFunction<DataFormat> = defaultTransformFunction,
    protected validateRecord: ValidateFunction<DataFormat> = defaultValidateFunction,
    protected updateRecord: UpdateFunction<DocumentData> = defaultUpdateFunction
  ) {
    super(undefined, observable.struct);

    // note that dataSource is not observable by default.
    // in the base case there is really no need for it
    makeObservable<this, "updateData">(this, {
      updateData: action,
    });
  }

  /**
   * Stores data on the observable `this.data` property. Mainly intended
   * to be used by the callback for the listener created by `this.subscribe`.
   * Validates the record and does not set `this.data` property if the result
   * fails validation.
   */
  protected updateData(snapshot: DocumentSnapshot): void {
    try {
      const snapshotData = snapshot.data({ serverTimestamps: "estimate" });

      if (!snapshotData) {
        this.data = undefined;
        this.hydrationState = { status: "hydrated" };
        return;
      }

      this.updateRecord(snapshotData);

      const record = this.transformRecord(snapshotData);

      if (record !== undefined) this.validateRecord(record);

      this.data = record;
      this.hydrationState = { status: "hydrated" };
    } catch (e) {
      this.hydrationState = { status: "failed", error: castToError(e) };
      this.data = undefined;
      // don't log routine feature flag checks, but do log everything else
      if (!(e instanceof FeatureGateError)) {
        Sentry.captureException(e);
      }
    }
  }

  protected startSnapshotListener(): Unsubscribe | undefined {
    return onSnapshot(
      this.dataSource,
      (result) => this.updateData(result),
      (error) => {
        runInAction(() => {
          this.hydrationState = { status: "failed", error };
        });
        Sentry.captureException(error);
      }
    );
  }
}
