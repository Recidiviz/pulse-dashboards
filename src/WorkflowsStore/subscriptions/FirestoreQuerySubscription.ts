// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import * as Sentry from "@sentry/react";
import {
  DocumentData,
  onSnapshot,
  Query,
  QuerySnapshot,
  Unsubscribe,
} from "firebase/firestore";
import {
  action,
  computed,
  makeObservable,
  observable,
  onBecomeObserved,
  onBecomeUnobserved,
  reaction,
  runInAction,
} from "mobx";
import { IDisposer } from "mobx-utils";

import {
  QuerySubscription,
  TransformFunction,
  ValidateFunction,
} from "./types";
import { defaultTransformFunction, defaultValidateFunction } from "./utils";

export abstract class FirestoreQuerySubscription<
  DataFormat extends DocumentData
> implements QuerySubscription<DataFormat>
{
  data: DataFormat[] = [];

  abstract get dataSource(): Query | undefined;

  cancelSnapshotListener?: Unsubscribe;

  disposeDynamicDataSource?: IDisposer;

  transformRecord: TransformFunction<DataFormat>;

  validateRecord: ValidateFunction<DataFormat>;

  constructor(
    transformFunction: TransformFunction<DataFormat> = defaultTransformFunction,
    validateFunction: ValidateFunction<DataFormat> = defaultValidateFunction
  ) {
    makeObservable<this, "updateData">(this, {
      data: observable,
      dataSource: computed,
      updateData: action,
      isLoading: observable,
      isHydrated: observable,
      error: observable,
      subscribe: action,
      resetHydration: action,
    });

    this.transformRecord = transformFunction;

    this.validateRecord = validateFunction;

    onBecomeObserved(this, "data", () => this.subscribe());
    onBecomeUnobserved(this, "data", () => this.unsubscribe());
  }

  protected updateData(snapshot: QuerySnapshot | undefined): void {
    const docs: DataFormat[] = [];

    const errors: unknown[] = [];

    snapshot?.forEach((doc) => {
      try {
        const record = this.transformRecord(
          doc.data({ serverTimestamps: "estimate" })
        );

        if (record !== undefined) this.validateRecord(record);

        if (record !== undefined) {
          docs.push(record);
        }
      } catch (e) {
        // Move on to allow for partial success
        errors.push(e);
        Sentry.captureException(e);
      }
    });

    this.data = docs;
    this.error = errors.length ? new AggregateError(errors) : undefined;

    this.isHydrated = true;
    this.isLoading = false;
  }

  get isActive(): boolean {
    return this.cancelSnapshotListener !== undefined;
  }

  subscribe(): void {
    // `subscribe`, on its own, will not re-subscribe if a listener already exists;
    // replacing the subscription with a new one requires further orchestration,
    // as seen in the reaction defined below.
    if (this.isActive || this.isLoading) return;

    this.isLoading = true;

    if (!this.dataSource) {
      this.updateData(undefined);
    } else {
      this.cancelSnapshotListener = onSnapshot(
        this.dataSource,
        (result) => this.updateData(result),
        (error) => {
          runInAction(() => {
            // an error that occurs after data is fetched is considered a partial failure,
            // but this one means that the fetch operation itself failed and therefore so does hydration.
            this.error = error;
            this.isHydrated = false;
            this.isLoading = false;
            this.data = [];
          });
          Sentry.captureException(error);
        }
      );
    }

    // watches dataSource for changes and re-subscribes accordingly
    this.disposeDynamicDataSource = reaction(
      () => this.dataSource,
      () => {
        this.unsubscribe();
        this.resetHydration();
        this.subscribe();
      }
    );
  }

  unsubscribe(): void {
    this.cancelSnapshotListener?.();
    this.cancelSnapshotListener = undefined;
    this.disposeDynamicDataSource?.();
  }

  hydrate(): void {
    this.subscribe();
  }

  isLoading: boolean | undefined = undefined;

  error: Error | undefined = undefined;

  isHydrated = false;

  resetHydration(): void {
    this.isLoading = false;
    this.isHydrated = false;
    this.error = undefined;
  }
}
