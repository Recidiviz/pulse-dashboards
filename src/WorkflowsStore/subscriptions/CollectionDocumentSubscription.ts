// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  doc,
  DocumentData,
  DocumentReference,
  DocumentSnapshot,
} from "firebase/firestore";
import { action, makeObservable, observable, override } from "mobx";

import { CollectionName, collectionNames, db } from "../../firestore";
import { FirestoreDocumentSubscription } from "./FirestoreDocumentSubscription";

/**
 * Subscribes to the specified document in the specified collection.
 * RecordType is essentially just an assertion for downstream convenience;
 * the data source itself is not validated or even strongly typed. Proceed with caution!
 */
export class CollectionDocumentSubscription<
  RecordType extends DocumentData
> extends FirestoreDocumentSubscription<RecordType> {
  readonly dataSource: DocumentReference<RecordType>;

  isLoading?: boolean;

  constructor(collectionName: CollectionName, recordId: string) {
    super();

    makeObservable(this, {
      isLoading: observable,
      subscribe: action,
      updateData: override,
    });

    this.dataSource = doc(
      db,
      collectionNames[collectionName],
      recordId
    ) as DocumentReference<RecordType>;
  }

  subscribe(): void {
    if (this.isLoading === undefined) {
      this.isLoading = true;
    }
    super.subscribe();
  }

  updateData(snapshot: DocumentSnapshot<RecordType>): void {
    super.updateData(snapshot);
    if (this.isLoading) {
      this.isLoading = false;
    }
  }
}
