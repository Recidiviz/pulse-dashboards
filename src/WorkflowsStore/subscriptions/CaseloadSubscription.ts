// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import {
  collection,
  DocumentData,
  Query,
  query,
  where,
} from "firebase/firestore";

import { CollectionName, collectionNames } from "../../FirestoreStore";
import { WorkflowsStore } from "../WorkflowsStore";
import { FirestoreQuerySubscription } from "./FirestoreQuerySubscription";

export class CaseloadSubscription<
  RecordType extends DocumentData
> extends FirestoreQuerySubscription<RecordType> {
  workflowsStore: WorkflowsStore;

  collectionId: string;

  personType: RecordType["personType"];

  constructor(
    workflowsStore: WorkflowsStore,
    collectionKey: CollectionName,
    personType: RecordType["personType"]
  ) {
    super();
    this.workflowsStore = workflowsStore;
    this.collectionId = collectionNames[collectionKey];
    this.personType = personType;
  }

  get dataSource(): Query | undefined {
    const {
      selectedOfficerIds,
      rootStore: { currentTenantId, firestoreStore },
    } = this.workflowsStore;

    if (!currentTenantId || !selectedOfficerIds.length) {
      return undefined;
    }

    const { collectionId, personType } = this;
    return query(
      collection(firestoreStore.db, collectionId),
      where("stateCode", "==", currentTenantId),
      where("officerId", "in", selectedOfficerIds)
    ).withConverter({
      fromFirestore(snapshot, options) {
        const doc = snapshot.data(options);
        return { ...doc, recordId: snapshot.id, personType };
      },
      // these collections are read-only, so this should never be used, but it is required by Firestore
      toFirestore(record: any) {
        throw new Error(`Writing to ${collectionId} is not allowed`);
      },
    });
  }
}
