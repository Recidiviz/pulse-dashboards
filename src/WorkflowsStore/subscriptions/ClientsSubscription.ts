// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { collection, Query, query, where } from "firebase/firestore";

import { ClientRecord, collectionNames, db } from "../../firestore";
import { WorkflowsStore } from "../WorkflowsStore";
import { FirestoreQuerySubscription } from "./FirestoreQuerySubscription";

export class ClientsSubscription extends FirestoreQuerySubscription<ClientRecord> {
  workflowsStore: WorkflowsStore;

  constructor(workflowsStore: WorkflowsStore) {
    super();
    this.workflowsStore = workflowsStore;
  }

  get dataSource(): Query | undefined {
    const {
      selectedOfficerIds,
      rootStore: { currentTenantId },
    } = this.workflowsStore;

    if (!currentTenantId || !selectedOfficerIds.length) {
      return undefined;
    }

    return query(
      collection(db, collectionNames.clients),
      where("stateCode", "==", currentTenantId),
      where("officerId", "in", selectedOfficerIds)
    ).withConverter({
      fromFirestore(snapshot, options) {
        const doc = snapshot.data(options);
        return { ...doc, recordId: snapshot.id };
      },
      // these collections are read-only, so this should never be used, but it is required by Firestore
      toFirestore(record: any) {
        throw new Error(`Writing to ${collectionNames.clients} is not allowed`);
      },
    });
  }
}
