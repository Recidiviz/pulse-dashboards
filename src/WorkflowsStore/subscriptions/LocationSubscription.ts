// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { collection, Query, query, where } from "firebase/firestore";

import { LocationRecord } from "../../FirestoreStore";
import { FIRESTORE_COLLECTIONS_MAP } from "../../FirestoreStore/constants";
import { RootStore } from "../../RootStore";
import { FirestoreQuerySubscription } from "./FirestoreQuerySubscription";

export class LocationSubscription extends FirestoreQuerySubscription<LocationRecord> {
  private rootStore: RootStore;

  constructor(rootStore: RootStore) {
    super();
    this.rootStore = rootStore;
  }

  get dataSource(): Query {
    const { searchField } = this.rootStore.workflowsStore;

    const stateCode = this.rootStore.currentTenantId;

    const constraints = [
      where("stateCode", "==", stateCode),
      where("idType", "==", searchField),
    ];

    return query(
      collection(
        this.rootStore.firestoreStore.db,
        FIRESTORE_COLLECTIONS_MAP.locations
      ),
      ...constraints
    );
  }
}
