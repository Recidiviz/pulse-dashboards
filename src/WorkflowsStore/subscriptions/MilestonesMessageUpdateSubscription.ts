// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { startOfToday } from "date-fns";
import { doc, DocumentReference } from "firebase/firestore";

import FirestoreStore, {
  collectionNames,
  MilestonesMessage,
} from "../../FirestoreStore";
import { getMonthYearFromDate } from "../utils";
import { CollectionDocumentSubscription } from "./CollectionDocumentSubscription";

export class MilestonesMessageUpdateSubscription<
  RecordType extends MilestonesMessage
> extends CollectionDocumentSubscription<RecordType> {
  // needs to be redefined because it's read-only but otherwise unchanged
  readonly dataSource: DocumentReference<RecordType>;

  constructor(firestoreStore: FirestoreStore, recordId: string) {
    const collectionName = "clientUpdatesV2";
    super(firestoreStore, collectionName, recordId);
    const dateKey = getMonthYearFromDate(startOfToday());

    this.dataSource = doc(
      firestoreStore.db,
      collectionNames[collectionName],
      recordId,
      collectionNames.milestonesMessages,
      dateKey
    ) as DocumentReference<RecordType>;
  }
}
