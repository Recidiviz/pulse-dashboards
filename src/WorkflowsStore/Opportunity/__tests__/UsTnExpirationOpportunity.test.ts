// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { configure } from "mobx";
import tk from "timekeeper";

import { RootStore } from "../../../RootStore";
import { Client } from "../../Client";
import { DocumentSubscription } from "../../subscriptions";
import {
  UsTnExpirationEligibleClientRecord,
  UsTnExpirationReferralRecordFixture,
} from "../__fixtures__";
import {
  hydrateExpirationDateRequirementText,
  UsTnExpirationOpportunity,
} from "../UsTnExpirationOpportunity";

jest.mock("../../subscriptions");

let opp: UsTnExpirationOpportunity;
let client: Client;
let root: RootStore;
let referralSub: DocumentSubscription<any>;
let updatesSub: DocumentSubscription<any>;

function createTestUnit(
  clientRecord: typeof UsTnExpirationEligibleClientRecord
) {
  root = new RootStore();
  jest
    .spyOn(root.workflowsStore, "opportunityTypes", "get")
    .mockReturnValue(["usTnExpiration"]);
  client = new Client(clientRecord, root);

  const maybeOpportunity = client.potentialOpportunities.usTnExpiration;

  if (maybeOpportunity === undefined) {
    throw new Error("Unable to create opportunity instance");
  }

  opp = maybeOpportunity;
}

beforeEach(() => {
  // this lets us spy on observables, e.g. computed getters
  configure({ safeDescriptors: false });
  tk.freeze(new Date(2022, 7, 1));
});

afterEach(() => {
  jest.resetAllMocks();
  tk.reset();
  configure({ safeDescriptors: true });
});

describe("fully eligible", () => {
  beforeEach(() => {
    createTestUnit(UsTnExpirationEligibleClientRecord);

    referralSub = opp.referralSubscription;
    referralSub.isLoading = false;
    referralSub.data = UsTnExpirationReferralRecordFixture;

    updatesSub = opp.updatesSubscription;
    updatesSub.isLoading = false;
  });

  test("requirements almost met", () => {
    expect(opp.requirementsAlmostMet).toEqual([]);
  });

  test("requirements met", () => {
    expect(opp.requirementsMet).toMatchSnapshot();
  });

  test("hydrate same day", () => {
    tk.reset();
    tk.freeze(new Date(2022, 1, 2, 16, 30));

    const actual = hydrateExpirationDateRequirementText(
      UsTnExpirationReferralRecordFixture.eligibleCriteria
        .supervisionPastFullTermCompletionDateOrUpcoming1Day
    );

    expect(actual).toEqual("Expiration date is today (Feb 2, 2022)");
  });

  test("hydrate text for 1 day past", () => {
    tk.reset();
    tk.freeze(new Date(2022, 1, 3, 16, 30));

    const actual = hydrateExpirationDateRequirementText(
      UsTnExpirationReferralRecordFixture.eligibleCriteria
        .supervisionPastFullTermCompletionDateOrUpcoming1Day
    );

    expect(actual).toEqual("1 day past expiration date (Feb 2, 2022)");
  });
});
