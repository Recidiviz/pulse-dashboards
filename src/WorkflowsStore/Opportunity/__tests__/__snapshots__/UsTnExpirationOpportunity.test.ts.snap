// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fully eligible requirements met 1`] = `
Array [
  Object {
    "text": "Expiration date is Feb 2, 2022",
    "tooltip": undefined,
  },
  Object {
    "text": "No zero tolerance codes since most recent sentence effective date",
  },
  Object {
    "text": "Not on lifetime supervision or lifetime sentence",
  },
]
`;
